### PR TITLE
This branch implements the CSV and native HA support for Hyper-V

### PR DESCRIPTION
**Pull request by anshul1886**
For design related details refer to https://cwiki.apache.org/confluence/display/CLOUDSTACK/iSCSI+and+HA+support+in+Hyper-V.

This is similar to VmWare in design and implementation.

### CODE REVIEW

**Comment by remibergsma** [comment]
Hi @anshul1886, 
Thanks for the PR :-) Looks like quite some work!
I don't have a Hyper-V setup to test it.. Will ask around and see who can help test it.

**Comment by runseb** [comment]
@anshul1886 looks like it failed the tests. Can you check what's happening and correct accordingly ?

**Comment by anshul1886** [comment]
@runseb These are false failures. The changes on which these tests are failing are in Hyper-V agent so can be safely ignored as these tests are not testing Hyper-V Agent.

**Comment by devdeep** [comment]
I'll take a look and get back if I have any queries

**Comment by anshul1886** [comment]
Rebased the code to latest master

**Comment by remibergsma** [comment]
@anshul1886 Can you update the PR? It doesn't merge clean.
Anyone able to review Hyper-v?

**Comment by anshul1886** [comment]
@remibergsma,
 Rebased the branch to latest master

**Comment by anshul1886** [comment]
Removed the MigrateCompleteCommand as per review comments

**Comment by anshul1886** [comment]
Unit test which is failing is testing KVM code and is not related to any changes in this file

**Comment by anshul1886** [comment]
Updated the code as per review comments

**Comment by devdeep** [comment]
Changes looks good to me. @remibergsma Are you ok with these changes? I'll go ahead and merge them then.

**Comment by remibergsma** [comment]
@devdeep Did you LGTM on this? Were you able to test / verify it with a hyper-v cluster?

@anshul1886 Can you look at the commits and see if you can squash some of them? I'm not saying it should be one commit, but I do want them to be atomic. Also, I don't see any tests for this? How are we going to make sure this keeps working once it is merged?

Pinging @karuturi @DaanHoogland to see if they can review the code as I'm not a Java guru. Thanks!

**Comment by anshul1886** [comment]
@remibergsma Either we have to squash all commits to one commit or leave it as it is. In squashing all commits to one I will put all the relevant messages. These commits don't have in order dependency. So what would you prefer one commit or leave it as it is?

Implementation have very little impact in core CloudStack so will not affect functionality of other Hypervisors. Also I have done testing on my end to make sure everything works fine after these commits.

Existing unit tests will be able to identify the issues in management server code.

**Comment by remibergsma** [comment]
@anshul1886 Then I'd say squash them into one so you commit the whole functionality as one commit. Also, when you add new code, I'd expect new unit tests as well? Otherwise the test coverage of ACS goes down.

File: engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -2294,7 +2292,7 @@ private void orchestrateMigrateWithStorage(final String vmUuid, final long srcHo
                 s_logger.info("Migration was unsuccessful.  Cleaning up: " + vm);
                 _alertMgr.sendAlert(alertType, srcHost.getDataCenterId(), srcHost.getPodId(),
                         "Unable to migrate vm " + vm.getInstanceName() + " from host " + srcHost.getName() + " in zone " + dc.getName() + " and pod " + dc.getName(),
-                        "Migrate Command failed.  Please check logs.");
+                        "Migrate Command failed. HA may not work for Hyper-V VM.  Please check logs.");
```
Comment (DaanHoogland) [inline engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java]: A hypervisor specific message in a generic module can lead to confusion. I would like to see the Hyper-V text replaced by a reference to hypervisortype or replaced by something more generic altogether

File: plugins/hypervisors/hyperv/DotNet/ServerResource/HypervResource/HypervResourceController.cs [hunk 2]
```diff
@@ -978,14 +1025,58 @@ public JContainer ModifyStoragePoolCommand([FromBody]dynamic cmd)
                     else if (poolType == StoragePoolType.NetworkFilesystem ||
                         poolType == StoragePoolType.SMB)
                     {
-                        NFSTO share = new NFSTO();
-                        String uriStr = "cifs://" + (string)cmd.pool.host + (string)cmd.pool.path;
-                        share.uri = new Uri(uriStr);
-                        hostPath = Utils.NormalizePath(share.UncPath);
+                        if ((bool)cmd.add)
```
Comment (DaanHoogland) [inline plugins/hypervisors/hyperv/DotNet/ServerResource/HypervResource/HypervResourceController.cs]: this ethod is getting very big. extracting some of the code to a private method would make it more readable (and probably more maintainable.

**Comment by DaanHoogland** [comment]
I don't see any test, we can not release this to the public without limitation of warranty. extensive testing needs to be implemented to give any degree of certainty on the functionality. there is not a single unit test and no reference to applicable integration tests or #!human scripts.

**Comment by anshul1886** [comment]
@DaanHoogland Updated the code with review comments.

Regarding unit tests I am identifying modules for which unit tests can be written. But be aware that unit tests are meant for development not for functional testing.

**Comment by DaanHoogland** [comment]
@anshul1886 if the purpose is to ensure that the path returned doesn't change, this seems perfectly aceptable to me. If people want to change the path in the future they must be aware and change the unit tests as well. That said they seem a litle trivial but I'm alright with that. It is mainly the integrity class hierarchy of the resources that is being tested this way.

**Comment by rhtyd** [comment]
@anshul1886 please re-base against latest master, update on current status of your work/PR

**Comment by anshul1886** [comment]
@swill @rhtyd Updated the PR against the latest master. It was conflicting in imports. Who is going to test this? AFAIK I am the only guy working in Hyper-V.

**Comment by swill** [comment]
Can you run the following tests on this PR and post the results?  I will run my KVM CI against it to verify nothing is broken for other hypervisors.

```
echo "Running tests with required_hardware=true"
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced,required_hardware=true \
smoke/test_password_server.py \
smoke/test_vpc_redundant.py \
smoke/test_routers_iptables_default_policy.py \
smoke/test_routers_network_ops.py \
smoke/test_vpc_router_nics.py \
smoke/test_router_dhcphosts.py \
smoke/test_loadbalance.py \
smoke/test_internal_lb.py \
smoke/test_ssvm.py \
smoke/test_vpc_vpn.py \
smoke/test_privategw_acl.py \
smoke/test_network.py

echo "Running tests with required_hardware=false"
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced,required_hardware=false \
smoke/test_routers.py \
smoke/test_network_acl.py \
smoke/test_reset_vm_on_reboot.py \
smoke/test_vm_life_cycle.py \
smoke/test_service_offerings.py \
smoke/test_network.py \
component/test_vpc_offerings.py \
component/test_vpc_routers.py
```

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 81
  Skipped: 0
   Failed: 0
   Errors: 1
 Duration: 8h 06m 18s
```

**Summary of the problem(s):**
```
ERROR: test suite for <class 'integration.smoke.test_internal_lb.TestInternalLb'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_internal_lb.py", line 296, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_XMQTJ1/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_17_2016_19_58_42_BEN56X:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/DeployDataCenter__May_17_2016_19_58_42_BEN56X/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/DeployDataCenter__May_17_2016_19_58_42_BEN56X/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/DeployDataCenter__May_17_2016_19_58_42_BEN56X/runinfo.txt)

**`/tmp/MarvinLogs/test_network_XMQTJ1:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/test_network_XMQTJ1/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/test_network_XMQTJ1/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/test_network_XMQTJ1/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_QO25FZ:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/test_vpc_routers_QO25FZ/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/test_vpc_routers_QO25FZ/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR351/tmp/MarvinLogs/test_vpc_routers_QO25FZ/runinfo.txt)


Uploads will be available until `2016-07-18 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
I have seen this error with other PRs.  I am not sure exactly where it comes from.  Out of curiosity, is the PR up-to-date in terms of rebasing?

I think this is a clean CI result considering I am seeing this same issue in some other PRs.

**Comment by anshul1886** [comment]
@swill Yes, PR is up to date. It was updated yesterday.

**Comment by swill** [comment]
Ok thanks.  I just wanted to verify.  I don't think I can merge this without integration tests against HyperV being run.  My tests just show it does not break other stuff, but we still don't have any proof it works for HyperV at this point.

**Comment by anshul1886** [comment]
@swill @rhtyd I am facing permission denied exceptions for some APIs in UI. Seems like they are after the merge of PR #1489.

**Comment by swill** [comment]
@rhtyd can you work with @anshul1886 to understand what is going on here?  This is important as we are about to freeze master and cut 4.9.  Thanks...

**Comment by rhtyd** [comment]
@swill yeah I've asked @anshul1886 few questions and waiting for answers. Btw, since we're in a release/branch freeze, this feature PR need to be held until 4.9.0 is released.

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 142
 Hypervisor xenserver
 NetworkType Advanced
 Passed=70
 Failed=3
 Skipped=3

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_vpc_vpn.py

 * test_01_vpc_remote_access_vpn Failed

* test_volumes.py

 * test_06_download_detached_volume Failed

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by jburwell** [comment]
@anshul1886 @rhtyd it appears that there was a push to get this merged for the 4.9 release, but a conflict may have occurred with the Dynamic Role-based API Checker enhancement.  Were you able to resolve the issue?  It also appears that @bvbharat had 3 test failures and 3 test skips.  Where does resolution of these test issues stand?

**Comment by rhtyd** [comment]
@anshul1886 can you rebase against latest master and push -f ?

**Comment by anshul1886** [comment]
Rebased the branch against current master.

**Comment by kiwiflyer** [comment]
tag:needslove

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 383
 Hypervisor xenserver
 NetworkType Advanced
 Passed=105
 Failed=0
 Skipped=7

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**

**Skipped tests:**
test_01_test_vm_volume_snapshot
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_routers_network_ops.py
test_disk_offerings.py

**Comment by DaanHoogland** [comment]
@kiwiflyer why does this need TLC? or is that remark deprecated?

**Comment by kiwiflyer** [comment]
It had been untouched for a while, so now that it's moving forward again, you can consider this tag depreciated.

**Comment by kiwiflyer** [comment]
@karuturi Could you see if you could get some folks from Accelerate to help review this PR? We'd really like to get this one moved forward. There are some comments above regarding some requested unit testing to make sure this doesn't break functionality. Help would be appreciated!

**Comment by haodeon** [comment]
Hi there, I found this PR after listening to a recording of Sowmya's presentation "Hyper-V Clustering with CloudStack".

The [Accelerite Blog](https://accelerite.com/blogs/hyper-v-clustering-enabling-ha-over-your-cloud-infrastructure/) implies CSV's and Failover Clustering is enabled in CloudStack but when I built a test environment I could not select PreSetup when adding Primary Storage as per the presentation slides.

It appears to me the presentation was based on functionality provided in this PR, is there any chance of this ever getting merged?

**Comment by DaanHoogland** [comment]
@anshul1886 please rebase and re-open if still relevant

**Pull request closed without merging**